pkg/statedb2/db.go

```diff
@@ -0,0 +1,219 @@
+// Copyright Authors of Cilium
+
+package statedb2
+
+import (
+	"errors"
+	"fmt"
+	"runtime"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	iradix "github.com/hashicorp/go-immutable-radix/v2"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/cilium/cilium/pkg/hive"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// DB provides an in-memory transaction database built on top of immutable radix
+// trees. The database supports multiple tables, each with one or more user-defined
+// indexes. Readers can access the data locklessly with a simple atomic pointer read
+// to obtain a snapshot. On writes to the database table-level locks are acquired
+// on target tables and on write transaction commit a root lock is taken to swap
+// in the new root with the modified tables.
+//
+// As data is stored in immutable data structures any objects inserted into
+// it MUST NOT be mutated afterwards.
+//
+// DB holds the "root" tree of tables with each table holding a tree of indexes:
+//
+//	           root
+//	          /    \
+//	         ba    T(foo)
+//	       /   \
+//	   T(bar)  T(baz)
+//
+//	      T(bar).indexes
+//		   /  \
+//		  i    I(byRevision)
+//		/   \
+//	   I(id)    I(ip)
+//
+//	          I(ip)
+//	          /  \
+//	        192  172
+//	        /     ...
+//	    bar(192.168.1.1)
+//
+// T = tableEntry
+// I = indexTree
+//
+// To lookup:
+//  1. Create a read (or write) transaction
+//  2. Find the table from the root tree
+//  3. Find the index from the table's index tree
+//  4. Find the object from the index
+//
+// To insert:
+//  1. Create write transaction against the target table
+//  2. Find the table from the root tree
+//  3. Create/reuse write transaction on primary index
+//  4. Insert/replace the object into primary index
+//  5. Create/reuse write transaction on revision index
+//  6. If old object existed, remove from revision index
+//  7. If old object existed, remove from graveyard
+//  8. Update each secondary index
+//  9. Commit transaction by committing each index to
+//     the table and then committing table to the root.
+//     Swap the root atomic pointer to new root and
+//     notify by closing channels of all modified nodes.
+//
+// To observe deletions:
+//  1. Create write transaction against the target table
+//  2. Create new delete tracker and add it to the table
+//  3. Commit the write transaction to update the table
+//     with the new delete tracker
+//  4. Query the graveyard by revision, starting from the
+//     revision of the write transaction at which it was
+//     created.
+//  5. For each successfully processed deletion, mark the
+//     revision to set low watermark for garbage collection.
+//  6. Periodically garbage collect the graveyard by finding
+//     the lowest revision of all delete trackers.
+type DB struct {
+	tables    map[TableName]TableMeta
+	mu        lock.Mutex // sequences modifications to the root tree
+	root      atomic.Pointer[iradix.Tree[tableEntry]]
```

Comment:
Lookup performance can be significantly improved by switching to a flat array presentation for the tables and indexes, e.g. assign each table an integer id, and each index an integer id within the table:
```
type root struct {
  tables []tableEntry
}
type tableEntry struct {
  indexes []iradix.Tree[object]
  revision Revision
}

root := db.root.Load()
index := root.tables[tablePos].indexes[indexPos]
```

I'll leave this optimization for a follow-up PR for now though.